egs/tedlium3/ASR/transducer_stateless/train.py

```diff
@@ -0,0 +1,728 @@
+        The return value of :func:`get_params`.
+      model:
+        The training model.
+      optimizer:
+        The optimizer that we are using.
+      scheduler:
+        The learning rate scheduler we are using.
+    Returns:
+      Return None.
+    """
+    if params.start_epoch <= 0:
+        return
+
+    filename = params.exp_dir / f"epoch-{params.start_epoch-1}.pt"
+    saved_params = load_checkpoint(
+        filename,
+        model=model,
+        optimizer=optimizer,
+        scheduler=scheduler,
+    )
+
+    keys = [
+        "best_train_epoch",
+        "best_valid_epoch",
+        "batch_idx_train",
+        "best_train_loss",
+        "best_valid_loss",
+    ]
+    for k in keys:
+        params[k] = saved_params[k]
+
+    return saved_params
+
+
+def save_checkpoint(
+    params: AttributeDict,
+    model: nn.Module,
+    optimizer: Optional[torch.optim.Optimizer] = None,
+    scheduler: Optional[torch.optim.lr_scheduler._LRScheduler] = None,
+    rank: int = 0,
+) -> None:
+    """Save model, optimizer, scheduler and training stats to file.
+
+    Args:
+      params:
+        It is returned by :func:`get_params`.
+      model:
+        The training model.
+    """
+    if rank != 0:
+        return
+    filename = params.exp_dir / f"epoch-{params.cur_epoch}.pt"
+    save_checkpoint_impl(
+        filename=filename,
+        model=model,
+        params=params,
+        optimizer=optimizer,
+        scheduler=scheduler,
+        rank=rank,
+    )
+
+    if params.best_train_epoch == params.cur_epoch:
+        best_train_filename = params.exp_dir / "best-train-loss.pt"
+        copyfile(src=filename, dst=best_train_filename)
+
+    if params.best_valid_epoch == params.cur_epoch:
+        best_valid_filename = params.exp_dir / "best-valid-loss.pt"
+        copyfile(src=filename, dst=best_valid_filename)
+
+
+def compute_loss(
+    params: AttributeDict,
+    model: nn.Module,
+    sp: spm.SentencePieceProcessor,
+    batch: dict,
+    is_training: bool,
+) -> Tuple[Tensor, MetricsTracker]:
+    """
+    Compute CTC loss given the model and its inputs.
+
+    Args:
+      params:
+        Parameters for training. See :func:`get_params`.
+      model:
+        The model for training. It is an instance of Conformer in our case.
+      batch:
+        A batch of data. See `lhotse.dataset.K2SpeechRecognitionDataset()`
+        for the content in it.
+      is_training:
+        True for training. False for validation. When it is True, this
+        function enables autograd during computation; when it is False, it
+        disables autograd.
+    """
+    device = model.device
+    feature = batch["inputs"]
+    # at entry, feature is (N, T, C)
+    assert feature.ndim == 3
+    feature = feature.to(device)
+
+    supervisions = batch["supervisions"]
+    feature_lens = supervisions["num_frames"].to(device)[: feature.size(0)]
+
+    texts = batch["supervisions"]["text"][: feature.size(0)]
+    y = sp.encode(texts, out_type=int)
```

Comment:
Please also handle `<unk>` in `texts`. Otherwise, `<unk>` would be encoded to 3 tokens here.

Reply:
I use the generated bpe.model (which is removed `<unk>`) to have a test. You mean this following case?
```
>>> import sentencepiece as spm
>>> sp = spm.SentencePieceProcessor()
>>> sp.load('data/lang_bpe_500/bpe.model')
True
>>> texts= ['<unk> imagine a place where your neighbors greet your children by name']
>>> y = sp.encode(texts, out_type=int)
>>> y
[[15, 2, 134, 41, 2, 17, 224, 35, 252, 6, 374, 248, 179, 324, 293, 67, 37, 42, 3, 65, 32, 169, 179, 450, 167, 15, 11, 19, 140]]
>>> y = sp.encode(texts, out_type=str)
>>> y
[['▁', '<', 'un', 'k', '>', '▁i', 'ma', 'g', 'ine', '▁a', '▁place', '▁where', '▁your', '▁ne', 'ig', 'h', 'b', 'or', 's', '▁g', 're', 'et', '▁your', '▁children', '▁by', '▁', 'n', 'a', 'me']]
>>>

```

Reply:
>[['▁', '<', 'un', 'k', '>', '▁i', 'ma', 'g', 'ine', '▁a', '▁place', '▁where', '▁your', '▁ne', 'ig'

You should handle the `<unk>` word separately. When you encounter an `<unk>` word in the transcript, please use
`sp.unk_id()` as its encoded result. 

Don't pass `<unk>` to `sp.encode()` directly.

Reply:
> I have run it again and will hava a test.

Please don't start your experiment until you fix the above comments.

Reply:
OK.